### PR TITLE
add mx quant

### PR DESCRIPTION
**Pull request by mengniwang95**
## Type of Change

feature

## Description

support mx quant

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

UT

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: neural_compressor/torch/algorithms/mx_quant/mx.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Tuple
+
+from neural_compressor.common.utility import MX_QUANT
+from neural_compressor.torch.quantization.config import MXQuantConfig
+from neural_compressor.torch.utils import register_algo, set_module
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/mx_quant/mx.py:25]: It seems that some `import` are not used.

File: examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        if task_name == "wikitext":
+            acc = results["results"][task_name]["word_perplexity"]
+        else:
+            acc = results["results"][task_name]["acc"]
+    print("Accuracy: %.5f" % acc)
+    print('Batch size = %d' % args.batch_size)
+
+if args.performance:
+    user_model.eval()
+    from intel_extension_for_transformers.llm.evaluation.lm_eval import evaluate
```
Comment (xin3he) [inline examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py:88]: How about adding a `requirement.txt` file?

**Review by xin3he — APPROVED** [review]
Could you please add a document to introduce the MX workflow and demo the usage?

**Comment by chensuyue** [comment]
@mengniwang95 any update?

**Comment by chensuyue** [comment]
Does this feature target v2.6?

File: neural_compressor/torch/algorithms/mx_quant/mx.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+from neural_compressor.torch.utils import set_module
+
+from .utils import quantize_elemwise_op, quantize_mx_op
+
+logger = Logger().get_logger()
+
+
+class MXLinearFunction(Function):
+    @staticmethod
+    def forward(ctx, input, weight, bias=None, mx_specs=None):
```
[the remark this answers is not in the document]
Comment (mengniwang95) [reply neural_compressor/torch/algorithms/mx_quant/mx.py:31]: mx_specs is MXQuantConfig

[the remark this answers is not in the document]
Comment (mengniwang95) [reply neural_compressor/torch/algorithms/mx_quant/mx.py:31]: mx_specs is got from configs_mapping.keys(), if other algorithms don't use dataclasses.dataclass during to_config_mapping process, it should not be introduced to mx_quant?

**Review by yiliu30 — APPROVED** [review]

**Comment by anthony-lemurian** [comment]
Not able to run the example, here is what i did

File: examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py
Comment (anthony-lemurian) [inline examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py:1]: Not able to run the example, here is what i did in a fresh venv
git clone https://github.com/intel/neural-compressor.git
cd neural-compressor
gh pr checkout 1728
pip install -r requirements.txt
python setup.py install
cd /neural-compressor/examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx
pip install -r requirements.txt
python3 -u run_clm_no_trainer.py --model bigscience/bloom-560m --quantize --accuracy --tasks lambada_openai --w_dtype fp4 --woq

then i get the error
2024-05-09 13:28:52 [INFO][algorithm_entry.py:512] Quantize model with the mx quant algorithm.
2024-05-09 13:29:01 [INFO][run_clm_no_trainer.py:61] Quantization end.
Traceback (most recent call last):
File "/home/anthony/venv_pr1728/neural-compressor/examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py", line 66, in
from intel_extension_for_transformers.transformers.llm.evaluation.lm_eval import evaluate
File "/home/anthony/venv_pr1728/lib/python3.10/site-packages/intel_extension_for_transformers/transformers/llm/evaluation/lm_eval/init.py", line 17, in
from .accuracy import cli_evaluate as evaluate
File "/home/anthony/venv_pr1728/lib/python3.10/site-packages/intel_extension_for_transformers/transformers/llm/evaluation/lm_eval/accuracy.py", line 42, in
from intel_extension_for_transformers.transformers.llm.evaluation.lm_eval import evaluator
File "/home/anthony/venv_pr1728/lib/python3.10/site-packages/intel_extension_for_transformers/transformers/llm/evaluation/lm_eval/evaluator.py", line 29, in
import lm_eval.api.metrics
ModuleNotFoundError: No module named 'lm_eval.api'

Comment (mengniwang95) [reply examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py:1]: Thank you for your reminder, we will update the requirements.txt. Please install intel-extension-for-transformers >= 1.4.1 and lm-eval==0.4.2

Comment (anthony-lemurian) [reply examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/mx/run_clm_no_trainer.py:1]: worked!

**Pull request merged**